src/main/java/at/ac/tuwien/kr/alpha/Alpha.java

```diff
@@ -0,0 +1,114 @@
+package at.ac.tuwien.kr.alpha;
+
+import at.ac.tuwien.kr.alpha.common.AnswerSet;
+import at.ac.tuwien.kr.alpha.common.Program;
+import at.ac.tuwien.kr.alpha.common.predicates.ExternalEvaluable;
+import at.ac.tuwien.kr.alpha.common.predicates.ExternalNativePredicate;
+import at.ac.tuwien.kr.alpha.common.terms.ConstantTerm;
+import at.ac.tuwien.kr.alpha.grounder.Grounder;
+import at.ac.tuwien.kr.alpha.grounder.GrounderFactory;
+import at.ac.tuwien.kr.alpha.grounder.parser.ProgramParser;
+import at.ac.tuwien.kr.alpha.solver.Solver;
+import at.ac.tuwien.kr.alpha.solver.SolverFactory;
+import at.ac.tuwien.kr.alpha.solver.heuristics.BranchingHeuristicFactory;
+import org.reflections.Reflections;
+import org.reflections.scanners.MethodAnnotationsScanner;
+import org.reflections.util.ClasspathHelper;
+import org.reflections.util.ConfigurationBuilder;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class Alpha {
+	private final Map<String, at.ac.tuwien.kr.alpha.common.predicates.Predicate> predicateMethods = new HashMap<>();
+
+	private final String grounderName;
+	private final String solverName;
+	private final String storeName;
+
+	private Solver solver;
+
+	private Program program;
+
+	public long getSeed() {
+		return seed;
+	}
+
+	public void setSeed(long seed) {
+		this.seed = seed;
+	}
+
+	public void setDeterministic() {
+		setSeed(0);
+	}
+
+	private long seed;
+
+	public Alpha(String grounderName, String solverName, String storeName) {
+		this.grounderName = grounderName;
+		this.solverName = solverName;
+		this.storeName = storeName;
+	}
+
+	public Alpha(String grounderName, String solverName) {
+		this(grounderName, solverName, "alpharoaming");
+	}
+
+	public Alpha(String grounderName) {
+		this(grounderName, "default");
+	}
+
+	public Alpha() {
+		this("naive");
+	}
+
+	public void scan(String base) {
+		Reflections reflections = new Reflections(new ConfigurationBuilder()
+			.setUrls(ClasspathHelper.forPackage(base))
+			.setScanners(new MethodAnnotationsScanner())
+		);
+
+		Set<Method> predicateMethods = reflections.getMethodsAnnotatedWith(Predicate.class);
+
+		for (Method method : predicateMethods) {
+			this.register(method);
+		}
+	}
+
+	public void register(Method method) {
+		this.predicateMethods.put(method.getName(), new ExternalEvaluable(method));
+	}
+
+	public void register(String name, java.util.function.Predicate<ConstantTerm> predicate) {
+		this.predicateMethods.put(name, new ExternalNativePredicate(name, predicate));
+	}
```

Comment:
Yes. The `Predicate` lambda will be called whenever the truth of some atom over the predicate `"&" + name` is "queried". TBH I can't really think of any name that is much more "catchy". Do you have suggestions? Otherwise I'll just add some comment to clarify.

Reply:
I think it should be called `RegisterSomething` where `Something` is an appropriate name for what is registered here. At the moment, I am, however, really confused about the terminology. I tried to figure it out looking at classes in the package `alpha.common.predicate` but got even more confused. I think those class names should be cleaned up somehow. Currently, there is Evaluable, External, ExternalEvaluable (which does not implement the Evaluable interface?), and ExternalNativeEvaluable (which I guess is the implementation of an external in Java?). And then there is the `Method` from reflections used as a parameter above.

By the way, the `Evaluable` has a method `evaluate` and here you said the `Predicate` lambda will be used to decide the truth of an external atom. Should those two things not be the same or what is their difference?

Basically, what is needed for external atom evaluation is:
* an interface of one evaluation method that maps a list of k terms (input) to a list of l terms (output).
* an predicate class for external predicates which extends ordinary predicates with the information of how many input terms and how many output terms are there.
* a class storing the association of evaluation methods and external predicates (where evaluation methods are registered as above).

Maybe I am forgetting something now. I also assume those three types are already here, but maybe under a different name than I would expect and probably organized differently.

Reply:
OK, then I would go for `registerPredicate`.

My picture of these things is as follows:

We have predicates, which are just words/names and a corresponding arity. They are represented by `Predicate`.

We distinguish between internal and external predicates:

 * For internal predicates, their interpretation (e.g. set of n-tuples of terms that make an atom over the predicate true) is defined through rules in the program.
 * For external predicates, their interpretation is computed by some interpretation function which is not part of the ASP program.

We further distinguish external predicates:

 * For external predicates with fixed interpretation the interpretation function takes n ground terms and returns a boolean truth value.
 * For external predicates with varying interpretation, the interpretation function takes n ground terms, and the current truth assignment and returns a boolean truth value.

Now, the change set that we are talking about here, introduces `FixedEvaluable` as an interface that external predicates with fixed interpretation should implement. Now we have some possibilities to do that:

 * Write a Java class that implements `Predicate` and `FixedEvaluable`. Encode its interpretation function in the `evaluate` method.
 * Just associate a name and an existing method, and generate an object of some type that implements `Predicate` and `FixedEvaluable`. The class `ExternalEvaluable` serves this purpose.  When it is being evaluated, then it "delegates" and calls the `Method` that was given at construction time. I might rename this class to something more descriptive based on our discussion, maybe `ExternalMethodEvaluable`.
 * Just associate a name an existing function. Since Java 8 allows for anonymous functions (lambda expressions) I thought that we might also support these. The principle is the same as for method references and the family of classes that extend `BaseExternalNativePredicate` serve this purpose. Java comes with `Predicate<T>`, `BiPredicate<T, U>`, `IntPredicate` and `LongPredicate` (I ignored `DoublePredicate` for now). So for all of them we have a corresponding class, that wraps an object of one of these types and creates a new object of a type that again implements `FixedEvaluable` and `Predicate`.

> By the way, the `Evaluable` has a method `evaluate` and here you said the `Predicate` lambda will be used to decide the truth of an external atom. Should those two things not be the same or what is their difference?

Whenever `FixedEvaluable.evaluate` is called, depending on which of the three above routes were taken to construct the predicate, the following might happen:

 * The object is of a type that has a "concrete" implementaion for `evaluate`, this will just be executed.
 * The object was constructed with a method reference, then the call to `evaluate` is delegated to that method.
 * The object was constructed with a function, then `evaluate` is delegated to the `test` method of the predicate.

So, up to now I only have worked on making "fixed interpretations" possible. I thought that having this makes sense on it's own, but also it is simpler than directly implementing atoms that bind new variables.

The interface `Evaluable` with its method that takes a list of terms and returns a list of terms, is there to be used for what you are sketching. However, we currently have no implementing classes. I might as well remove that interface.

In order to implement atoms that bind variables, I think I am missing some puzzle pieces. Up until now, the changes on a "technology" level kept me busy, and actually I wanted to get some working prototype as fast as possible. What I still do not understand is how the truth value of these external atoms is defined. Suppose i have `employee(Y) :- employee(X), managerOf[Y](X)` (if X is an employee, and Y is the manager of X, then Y is also an employee), then how do I determine whether `manager[Y](X)` is true or false? Also, must these external atoms always map to one single term? What if `X`has more than one manager?
Probably this is all covered by the definition of external atoms. Could you refer me to some paper?

What still confuses me is why these constructions are called "external atom" and not "external predicate". This might however indicate a fundamental flaw in my thinking. For me, predicates that use external functions to define their interpretations can very well be called external predicates. However, only atoms may bind variables, so we then also get atoms that use functions to bind variables, which would be something else than an external predicate, namely an external atom.

Also, how does all this relate to function terms? Wouldn't an external atom that binds exactly one term have the same power as a function?

Reply:
Ok now there are several separate poins to discuss. Let me enumerate those issues for better reference:

1. Method name:
I think `registerPredicate` is an improvement over the current name, but if I understand correctly, then this method also registers the implementation of the predicate, so I would suggest `registerPredicateImplementation`.

2. Predicate notion:
A predicate is just a name and an arity, we agree on that. This is my reason why the method under 1. should not just be `registerPredicate` because it does more than just saying there exists a name with such arity.

3. Distinction internal and external predicates:
I basically agree but would like to add that internal predicates are usually defined by the rules of the ASP program but also may be builtin and have a fixed interpretation (like the `=` predicate, for example).

4. Should external atoms receive the current interpretation:
This is an important question and I thought that we decided on No. If we decide that external atoms may depend on the current interpretation, then we have a system similar to HEX, which is very powerful but also harder to evaluate. If the truth of the external atom depends on an interpretation, we have to re-evaluate the external atom every time the interpretation changes (conceptually, in practice some evaluations can be avoided, but in the worst case we have to). This then implies that evaluating external atoms cannot be done simply during the grounding but much more work is needed. In fact, something similar as in the `alphahex` branch has to happen where Alpha is paired with the HEX solver and quite some necessary work is done in HEX (like computing a full grounding of certain parts of the input program).
Also note that terms have a fixed interpretation (since ASP is based on Herbrand interpretations where each term is interpreted by itself). So for an external atom `&external( term1, f(term2) )` the truth of the whole atom may depend on the implementation of the external atom, but its input will always be the same, namely ( term1, f(term2) ) regardless of the current interpretation.
Summarizing, I strongly suggest that we only allow `FixedEvaluables`.

5. Output of external atoms:
I think an external atom can give us some output (maybe even a set of outputs) in the form or terms. So, an external atom is a mapping from `k` terms to a set of `l`terms. From the theoretical view this is simply a `k+l`-ary Boolean function. Evaluating an external atom with `k` input terms then results in a tuple (or set of tuples) of `l` ouput terms where the reading is that the `k+l`-ary Boolean function is true for (each combination of) the input terms and the output terms and false for all other combinations of the given input terms with any possible output terms.

6. Multiple ways to create the implementation of external atoms:
I agree that for usability it is good to have multiple ways to actually implement the external. At the moment, I do not understand how the anonymous functions should work. I think it should always be a function from `List<Term>` to `List<Term>` (or rather to `Set<List<Term>>`, see below). How come `BiPredicate`, `IntPredicate` and so on into play here?

7. The missing piece on the `Evaluable` interface:
A function `f` mapping `k` input terms to `l` output terms can be represented as a Boolean function `g` on `k+l` input terms as follows:
`f(in_1, ..., in_k) = (out_1, ..., out_l)` iff `g(in_1, ..., in_k, out_1, ..., out_l)` is true.
So, conceptually we use external atoms in the style of `g` but to evaluate them we use the implementation of `f`.

8. Variable binding:
I think this should be done purely by the grounder and not inside the external atoms. So the grounder will ensure that an external atom is called only after all its input terms are bound and take the return values of the external atom and bind variables at output positions accordingly.

9. On the concrete example:
`employee(Y) :- employee(X), managerOf[Y](X)`
First, we have to clear what is input and output. In HEX the part between `[ ]` is input and between `( )` is ouput -- I would stick to that. So, let me correct the rule to:
`employee(Y) :- employee(X), &managerOf[X](Y).`
Now, the grounder looks for bindings of `X` and for each binding of `X` it will evaluate the implementation of `&managerOf` and get in return a set of bindings for `Y`. So the type of our implementation should be `Set<List<Term>> method(List<Term>)`. So for each binding `X=a` we call `method(a)` and get a set `{ (b), (c), (d) }` of all managers of `a`; in this case `b`, `c` and `d` are all managers of `a`. Now the grounder can bind `Y`and it generates a new ground instance of this rule for every binding (so we get 3 ground instances in this example).

10. References:
* http://www.kr.tuwien.ac.at/staff/redl/publications/ekrsw2017-rw.pdf
This is a summer school article introducing HEX and its foundations. It also contains some small examples but does not go into the details of the implementation.
* https://www.ijcai.org/proceedings/2017/0141.pdf
This shows how one can combine HEX and lazy grounding using Alpha (where the truth value of external atoms depends also on the current assignment). Although the Algorithm is not far from Alpha, there are several notions where no algorithm is given but which must be computed. These notions are input-safe domain, external input assignment, external evaluation function, and on top of that, all implementations of external atoms must be assignment-monotonic. Hence, I strongly suggest to not let external atoms depend on the current assignment but only on the given term input.

11. I think we could use a concrete example for discussion and later for testing. What did you have in mind when starting with all of this?

12. Why it is called external atom and not external predicate:
Well, the predicate is just a name and some arity. In order to speak about truth, we need some terms (of correct arities), hence we get an atom.

13. Connection to function terms:
In general, an interpretation associates each function term with one element from the universe. So yes, what the external atom is doing could be seen as a just a certain interpretation of a function term. But in logic programming and ASP we use Herbrand interpretations and those have a fixed interpretation for constants and function terms. In an Herbrand interpretation the universe always is the set of all terms and each constant `c` is mapped to the element `c` of the universe. Likewise, every function term `f(t_1,...,t_n)` is mapped to the element `f(t_1,...,t_n)` of the universe. Jaques Herbrand could show that for every first order formula `F` it holds that `F` is satisfiable if-and-only-if there exists a Herbrand interpretation that satisfies `F`. In essence, this allows us to decide validity of a formula by only looking at its syntactic form. Now, one might be able to realize external atoms by allowing for certain function symbols to not be interpreted in Herbrand-style, but then I would expect that things get ugly.

14. I think this work you do is totally worth it and it is great to see that development happen!

Reply:
Thank your the thorough answer! I think we agree on most of the items, especially regarding (4.) I also do not want to go down the rabbit hole of passing around the current assignment/interpretation.

#### Regarding (6.):
> How come `BiPredicate`, `IntPredicate` and so on into play here?

Take for example the rule `reachable(A,C) :- &connected(A,B), &connected(B,C).` and further `:-
 connected(A,B), not node(A), not node(B)`, and assume that we have all objects that are nodes introduced to the program via facts. However, the question of answering whether two of these nodes are connected should be answered by a Java lambda expression of type `BiPredicate<Node, Node>`, that takes the two nodes, fiddles around with them in unexpected ways, and then gives us the answer whether two nodes (substituted for `A`, `B` and `C`) are connected. This comes very close to what you suggested in (5.), namely

> From the theoretical view this is simply a k+l-ary Boolean function.

So, here you have that function. In this special case k = l = 1. The difference to functions like you sketched in (9.), namely

> `Set<List<Term>> method(List<Term>)`

is that our `BiPredicate<Node,Node>` cannot return terms, so we cannot do variable binding as you suggest in (8.). I expect that these types of "interpretation functions" are easier to understand and implement, as they leave the assigning of variables/the binding for the grounder. Still, I think that even a simple use case as the one above is perfectly fine and makes sense to me.

This is specifically the reason why I implemented them first. I did not want to bother handling returned terms and binding them for a given rule.

#### Regarding (9.)
What would `Set<List<Term>> method(List<Term>)` return if there is no manager for some binding for `X`? The empty set? This way it should be relatively straight forward to implement.

Reply:
Thanks for the explanation regarding 6.  Allowing these as special cases seems okay.

Regarding 9. yes, exactly it would be the empty set.
In case of an external atom with no return value it would be a singleton set containing an empty list of terms, i.e., the term `()`, if the atom is true, and the empty set if the atom is false.